### PR TITLE
Add support for cross-compiling on iOS

### PR DESCRIPTION
**Pull request by freakboy3742**
[PEP 730](https://peps.python.org/pep-0730/) adds iOS as a Tier 3 supported platform for CPython 3.13. This PR adds the modifications to crossenv to support cross-compiling for iOS devices (ARM64) and iOS Simulators (ARM64 and x86_64). It also includes support for tvOS and watchOS, since the changes are effectively identical.

It also migrates the `setup.py` file to PEP 621 `pyproject.toml` format. A `pyproject.toml` file with a PEP517 configuration is required to avoid warnings (and, in the very near future installation errors); migrating the rest of the configuration was relatively trivial.

These changes have been empirically tested through use in [Mobile Forge](https://github.com/beeware/mobile-forge), which compiles a range of binary packages, for Python 3.9 through 3.13.

There's no additional *programmatic* tests, because it's not clear if the existing test suite is currently in a working state. I couldn't find any documentation for how to run the test suite; when I take the naïve approach and use `pytest` to run the test suite locally (on macOS), I get a 100% failure rate; and there's no CI results (that I can find) to compare against. I'm happy to add runtime tests for any of these changes if requested - I just need some guidance on how the test suite is *meant* to run.

### CODE REVIEW

**Comment by benfogle** [comment]
Wow, this is really neat, thank you. I will start taking a look at it right away.

The existing test suite is unfortunately, as you noticed, partially broken. Some of the tests ended up depending on a combination of Python/setuptools/3rd party modules, which has ended up being far too much of a moving target to maintain. This is something I need to rework.

**Comment by freakboy3742** [comment]
If you've got any general notes/ideas on how the test suite *should* work (or it's a case of "just get it working, however you do it), I might be able to find some time to work on fixing those issues. However, if you'd prefer to tackle that kind of "core infrastructure" yourself, I'd completely understand.

**Review by benfogle — CHANGES_REQUESTED** [review]
I'm still checking some of the tests manually to make sure nothing broke. I'm not able to verify the iOS side of things.

**Comment by benfogle** [comment]
Re: the test suite. It's not terribly broken and can be run manually. I do actually have a local branch to fix things up. Basically it needs:

* Migrate to GitHub actions
* Instead of maintaining a toolchain for testing, run the tests in pre-built docker images. There are docker images that would work that either didn't exist when I started the project, or I wasn't aware of them.
* Don't rely on building 3rd party modules for test purposes. Like bcrypt started using Rust and numpy started using cmake. And setuptools changes too. This is tricky, because testing with setuptools is still a good idea even though it's a moving target.

File: crossenv/__init__.py [hunk 1]
```diff
@@ -676,8 +733,9 @@ def make_cross_python(self, context):
         if self.cross_prefix:
             context.cross_env_dir = self.cross_prefix
         else:
-            context.cross_env_dir = os.path.join(context.env_dir, 'cross')
-        clear_cross = self.clear in ('default', 'cross-only', 'both')
+            cross_env_name = os.path.split(context.env_dir)[-1]
+            context.cross_env_dir = os.path.join(context.env_dir, cross_env_name)
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply crossenv/__init__.py:737]: The reason is that "cross" isn't a helpful name in the context of iOS compilation. An iOS build is actually *3* separate cross-platform builds (arm64-iphoneos; arm64-iphonesimulator; and x86_64-iphonesimulator). Always using the name "cross" means you can't easily tell what platform you're currently compiling for, and removes any possibility of having cross-environment directories side-by-side.

File: crossenv/__init__.py [hunk 2]
```diff
@@ -515,10 +546,15 @@ def get_uname_info(self):
             if len(host_info) > 1 and host_info[-1] in platform2uname:
                 # Test that this is still a special case when we can.
                 self.host_machine = platform2uname[host_info[-1]]
+            elif self.host_sys_platform in {"ios", "tvos", "watchos"}:
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply crossenv/__init__.py:549]: Sure - CPython 3.13 calls that collection "apple mobile"; I'll add a constant to that effect.

**Comment by freakboy3742** [comment]
> Re: the test suite. It's not terribly broken and can be run manually. 

I've had no luck running pytest on the current master branch. I get 100% failures, with every test failing with either:
```
shutil.ReadError: /Users/rkm/projects/crossenv/tests/prebuilt/prebuilt_musl_arm_aarch64.tar.xz is not a compressed or uncompressed tar file
```
or
```
UnboundLocalError: cannot access local variable 'archives' where it is not associated with a value
```

Looking at the xz file - if I try to unpack it manually, I get an error that it isn't a valid archive "Unrecognised archive format".

File: crossenv/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply crossenv/__init__.py:740]: I guess this change isn't *technically* required; but I've found it *really* convenient and helpful. I guess I could walk this back if you feel the change is too invasive.

**Comment by freakboy3742** [comment]
I see now from reverse engineering the Azure config that the problem is likely the git lfs handling. I'll see if that fixes things...

File: crossenv/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benfogle) [reply crossenv/__init__.py:740]: Would the `prompt` param of [EnvBuilder](https://docs.python.org/3/library/venv.html#venv.EnvBuilder) work?

[the remark this answers is not in the document]
Comment (freakboy3742) [reply crossenv/__init__.py:740]: In what sense - using that name as an alternative suffix, or setting prompt when building the environment?

[the remark this answers is not in the document]
Comment (freakboy3742) [reply crossenv/__init__.py:740]: Oh - I see what you're saying now. Yes, that would work. Change incoming shortly.

**Comment by freakboy3742** [comment]
Ok - it's clearly not *just* lfs; I get binary format errors if I run on an ARM64 macOS laptop. From what I can work out, that's not surprising - the test (and lfs content) seems to involve packaged binaries. My x86_64 Ubuntu 22.04 test machine also fails - there's a handful of test successes (mostly in `test_usage`), but also a bunch of executable format errors, and a lot of "AssertionError: Could not make crossenv!".

Are you able to provide any high level context for what the test suite is actually doing, and where it *should* run?

**Comment by benfogle** [comment]
The test suites expect to run on Linux (crossenv has been Linux-only up to now), with qemu-user available via binfmt. On ubuntu, you can install `qemu-user-binfmt` or `qemu-user-static`. I run it on Ubuntu 22.04 currently.

Most of these tests will create a temporary crossenv environment (you can poke around in `/tmp/pytest-of-$USER` to see them) and then run commands in that environment. To make sure that the cross-compiled modules can execute, we use qemu-user to emulate the host-python and load the module. For debugging, I often find it helpful to debug a single test at a time: `pytest -k <name_of_test> --pdb`. You can also use `--trace` instead of `--pdb` to break at the start instead of on failure.

**Comment by freakboy3742** [comment]
> The test suites expect to run on Linux (crossenv has been Linux-only up to now), with qemu-user available via binfmt. On ubuntu, you can install `qemu-user-binfmt` or `qemu-user-static`. I run it on Ubuntu 22.04 currently.
> 
> Most of these tests will create a temporary crossenv environment (you can poke around in `/tmp/pytest-of-$USER` to see them) and then run commands in that environment. To make sure that the cross-compiled modules can execute, we use qemu-user to emulate the host-python and load the module. For debugging, I often find it helpful to debug a single test at a time: `pytest -k <name_of_test> --pdb`. You can also use `--trace` instead of `--pdb` to break at the start instead of on failure.

Ah - that probably explains it. My Ubuntu 22.04 test box is already being emulated, so QEMU won't work (I can't run the Android emulator on my Ubuntu or Windows test boxes for essentially the same reason).

Using a Docker container to run the tests suite container might avoid those problems - but it might hit the same "emulation inside an emulator" limits. I'll try and see if that's any more successful.

**Comment by freakboy3742** [comment]
@benfogle I've done some more investigation; it looks like the issue isn't QEMU as I thought, but the libraries dynamically linked by the prebuilt environments. Details in #118 where I've got a GitHub Actions CI configuration definition.

**Comment by benfogle** [comment]
I'm not seeing any breakage manually. What I'm going to do is cut a release off of another PR that just went in and then merge this one. If there is breakage we're not catching, it'll be on the unreleased master branch and hopefully we can fix up CI and then release.

**Pull request merged**